File: python/08_062.py
def numIslands(grid):

    if not grid or not grid[0]:
        return 0
    
    rows, cols = len(grid), len(grid[0])
    count = 0
    
    def dfs(i, j):
        if (i < 0 or i >= rows or j < 0 or j >= cols or
            grid[i][j] == '0' or grid[i][j] == '2'):
            return
        
        # Mark as visited
        grid[i][j] = '0'
        
        # Explore all four directions
        dfs(i+1, j)
        dfs(i-1, j)
        dfs(i, j+1)
        dfs(i, j-1)
    
    for i in range(rows):
        for j in range(cols):
            if grid[i][j] == '1':
                count += 1
                dfs(i, j)
    
    return count

File: cpp/08_062.cpp
#include <vector>
using namespace std;

class Solution {
public:
    // Count number of islands in a grid
    int numIslands(vector<vector<char>>& grid) {
        // Check if grid is empty
        if (grid.empty() || grid[0].empty()) {
            return 0;
        }
        
        int rows = grid.size();
        int cols = grid[0].size();
        int count = 0;
        
        // Traverse the grid
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                // If land is found, run DFS to mark island
                if (grid[i][j] == '1') {
                    count++;
                    dfs(grid, i, j, rows, cols);
                }
            }
        }
        
        return count;
    }
    
private:
    // Run DFS to mark connected land cells
    void dfs(vector<vector<char>>& grid, int i, int j, 
            int rows, int cols) {
        // Check boundaries and if cell is water or visited
        if (i < 0 || i >= rows || j < 0 || j >= cols || 
            grid[i][j] == '0') {
            return;
        }
        
        // Mark as visited
        grid[i][j] = '0';
        
        // Explore all four directions
        dfs(grid, i+1, j, rows, cols);
        dfs(grid, i-1, j, rows, cols);
        dfs(grid, i, j+1, rows, cols);
        dfs(grid, i, j-1, rows, cols);
    }
};